truffle/src/main/java/org/jruby/truffle/language/SnippetNode.java

```diff
@@ -67,13 +68,25 @@ public Object execute(VirtualFrame frame, Object... arguments) {
                     true,
                     this);
 
-            directCallNode = insert(Truffle.getRuntime().createDirectCallNode(Truffle.getRuntime().createCallTarget(rootNode)));
+            directCallNode = insert(Truffle.getRuntime().createDirectCallNode(
+                    Truffle.getRuntime().createCallTarget(rootNode)));
 
             if (directCallNode.isInlinable()) {
                 directCallNode.forceInlining();
             }
         }
 
+        if (arguments.length != parameters.length * 2) {
+            CompilerDirectives.transferToInterpreter();
+            throw new UnsupportedOperationException(
+                    "number of arguments doesn't match number of parameters");
+        }
+
+        assert this.expression.equals(expression);
```

Comment:
I was thinking about cases where some expression could be shared between similar specialisations with possibly some small changes. It seemed not important in assert.